### PR TITLE
code-gen: Simplify `switch`

### PR DESCRIPTION
**Pull request by dweindl**
Improves model code generation by collapsing cases with identical statements.

I.e.
```
switch(a)
 case b:
 case c:
  statements;
  break;
```
instead of

```
switch(a)
 case b:
  statements;
  break;
 case c:
  statements;
  break;
```

For my current model of interest, containing many events, this significantly reduces the generated code:

E.g.:
```
 16K my_model/deltasx.cpp
6,6M my_model_old/deltasx.cpp
```

Overall, for this model, I got from 204201 LOC down to 7936 LOC (i.e. -96%).

### CODE REVIEW

File: python/sdist/amici/cxxcodeprinter.py [hunk 1]
```diff
@@ -303,7 +303,9 @@ def get_switch_statement(
     indentation_step: Optional[str] = " " * 4,
 ):
     """
-    Generate code for switch statement
+    Generate code for a C++ switch statement.
+
+    Generate code for a C++ switch statements with a ``break`` after each case.
```
Comment (dilpath) [inline python/sdist/amici/cxxcodeprinter.py:308]:
```suggestion
    Generate code for a C++ switch statement with a ``break`` after each case.
```

File: python/sdist/amici/cxxcodeprinter.py [hunk 2]
```diff
@@ -321,26 +323,43 @@ def get_switch_statement(
...
-
-    return lines
+            case_code = f"{indent1}case {expression}:"
+
+            try:
+                # there is already a case with the same statement, append
+                cases_map[statement_code].append(case_code)
+            except KeyError:
+                # add new case + statement
+                cases_map[statement_code] = [case_code]
```
Comment (dilpath) [inline python/sdist/amici/cxxcodeprinter.py:351]:
```suggestion
            cases_map[statement_code] = cases_map.get(statement_code, []) + [case_code]
```

File: python/sdist/amici/cxxcodeprinter.py [hunk 3]
```diff
@@ -321,26 +323,43 @@ def get_switch_statement(
...
+    def get_lines():
+        for statements, case_code in cases_map.items():
+            yield from case_code
+            yield from statements
+
+    return [
+        f"{indent0}switch({condition}) {{",
+        *(get_lines()),
+        indent0 + "}",
+    ]
```
Comment (dilpath) [inline python/sdist/amici/cxxcodeprinter.py:365]: Same or different?

```suggestion
    return [
        f"{indent0}switch({condition}) {{",
        *[code for k, v in cases_map.items() for code in [k,v]],
        indent0 + "}",
    ]
```

**Review by dilpath — COMMENTED** [review]
Nice :+1:

File: python/sdist/amici/cxxcodeprinter.py (hunk 3 above)
Comment (dweindl) [reply python/sdist/amici/cxxcodeprinter.py:365]: Same result, but constructs potentially plenty of temporary lists. Preferring the former.

Comment (dilpath) [reply python/sdist/amici/cxxcodeprinter.py:365]: Ah, how about

```suggestion
    return [
        f"{indent0}switch({condition}) {{",
        *[code for codes in cases_map.items() for code in codes],
        indent0 + "}",
    ]
```

**Review by FFroehlich — APPROVED** [review]
nice 👍

File: python/sdist/amici/cxxcodeprinter.py (hunk 3 above)
Comment (dweindl) [reply python/sdist/amici/cxxcodeprinter.py:365]: Would work with changing `in codes` to `in chain.from_iterable(codes)`.

File: python/sdist/amici/cxxcodeprinter.py (hunk 2 above)
Comment (dweindl) [reply python/sdist/amici/cxxcodeprinter.py:351]: creates an unnecessary list each time

**Pull request merged**